source/extensions/quic_listeners/quiche/envoy_quic_proof_verifier_base.cc

```diff
@@ -0,0 +1,61 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_proof_verifier_base.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+
+#include "openssl/ssl.h"
+#include "quiche/quic/core/crypto/certificate_view.h"
+#include "quiche/quic/core/crypto/crypto_protocol.h"
+#include "quiche/quic/core/quic_data_writer.h"
+
+namespace Envoy {
+namespace Quic {
+
+quic::QuicAsyncStatus EnvoyQuicProofVerifierBase::VerifyProof(
+    const std::string& hostname, const uint16_t port, const std::string& server_config,
+    quic::QuicTransportVersion /*quic_version*/, absl::string_view chlo_hash,
+    const std::vector<std::string>& certs, const std::string& cert_sct,
+    const std::string& signature, const quic::ProofVerifyContext* context,
+    std::string* error_details, std::unique_ptr<quic::ProofVerifyDetails>* details,
+    std::unique_ptr<quic::ProofVerifierCallback> callback) {
+  quic::QuicAsyncStatus res = VerifyCertChain(hostname, port, certs, "", cert_sct, context,
+                                              error_details, details, std::move(callback));
+  if (res == quic::QUIC_FAILURE) {
+    return quic::QUIC_FAILURE;
+  }
+  ASSERT(res != quic::QUIC_PENDING);
```

Comment:
I'd swap the order of signature verification and cert verification here. Two reasons: 1) signature verification is cheaper than cert verification, so if it fails, you can skip the more expensive step, and 2) you don't have to worry about the callback being called asynchronously (if its behavior ever changes in the future).